orca-keel/src/main/kotlin/com/netflix/spinnaker/orca/keel/task/PublishDeliveryConfigTask.kt

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright 2017 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.orca.keel.task
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.fasterxml.jackson.module.kotlin.convertValue
+import com.netflix.spinnaker.orca.ExecutionStatus
+import com.netflix.spinnaker.orca.KeelService
+import com.netflix.spinnaker.orca.RetryableTask
+import com.netflix.spinnaker.orca.TaskResult
+import com.netflix.spinnaker.orca.igor.ScmService
+import com.netflix.spinnaker.orca.pipeline.model.GitTrigger
+import com.netflix.spinnaker.orca.pipeline.model.Stage
+import org.slf4j.LoggerFactory
+import org.springframework.http.HttpStatus
+import org.springframework.stereotype.Component
+import retrofit.RetrofitError
+import java.util.concurrent.TimeUnit
+
+/**
+ * Task that retrieves a Managed Delivery config manifest from source control via igor, then publishes it to keel,
+ * to support GitOps flows.
+ */
+@Component
+class PublishDeliveryConfigTask
+constructor(
+  private val keelService: KeelService,
+  private val scmService: ScmService,
+  private val objectMapper: ObjectMapper
+) : RetryableTask {
+  private val log = LoggerFactory.getLogger(javaClass)
+
+  override fun execute(stage: Stage): TaskResult {
+    val context = objectMapper.convertValue<PublishDeliveryConfigContext>(stage.context)
+    val trigger = stage.execution.trigger
+    val user = trigger.user ?: "anonymous"
+
+    if (context.attempt > context.maxRetries!!) {
+      val error = "Maximum number of retries exceeded (${context.maxRetries})"
+      log.error("$error. Aborting.")
+      return TaskResult.builder(ExecutionStatus.TERMINAL).context(mapOf("error" to error)).build()
+    }
+
+    // if the pipeline has a git trigger, infer what context we can from the trigger
+    if (trigger is GitTrigger) {
+      // FIXME: GitTrigger props are non-null, but IIUC branch and hash would be mutually exclusive, so what to do here?
```

Comment:
Since the hash trumps the branch, for now I'm prioritizing the hash. Please let me know if you understand this differently.